### PR TITLE
[UI Tests] - Update add products tests

### PR DESCRIPTION
**Pull request by jostnes**
## Description
This PR updates the existing add product types UI tests. I realized after we were using only 1 mock file with `"type": "simple",` and validating only the product name on the published product screen. While this is correct for physical and virtual products, this still passed on variable test because it was only checking the product name even though the incorrect product screen was displayed on the UI. 

To fix this, I created a different mock file for each product type (`physical`, `virtual`, `variable`, `grouped` and `external`) that will be called for the different product types. I also renamed `verifyNewProductScreenLoaded` to `verifyPublishedProductScreenLoaded` and added a unique validation per screen, intentionally keeping it at one assertion type per type so that it doesn't take long but enough info that the correct page is displayed.

Note: A lot of file changes because of the mock files. Code changes (and updates to `WooCommerce/WooCommerce.xcodeproj/project.pbxproj`) are in https://github.com/woocommerce/woocommerce-ios/pull/9441/commits/3bf4e1a27c78c06b9ae19a9dd34e853c68714288, mock file additions are in https://github.com/woocommerce/woocommerce-ios/pull/9441/commits/efd3ac9b9922ec14d80f68245e810c88f33bbbe0 

## Testing instructions
Products-related tests should work as expected.

### CODE REVIEW

File: WooCommerce/WooCommerceUITests/Flows/MockDataReader.swift [hunk 1]
```diff
@@ -39,8 +39,14 @@ class GetMocks {
...
 
-    static func readNewProductData() throws -> ProductData {
-        let originalData = try JSONDecoder().decode(NewProductMock.self, from: self.getMockData(test: ProductsTests.self, filename: "products_add_new"))
+    static func readNewProductData(productType: String) throws -> ProductData {
+        let file = [
+            "physical": "products_add_new_physical_2129",
+            "virtual": "products_add_new_virtual_2123",
+            "variable": "products_add_new_variable_2131"
+        ]
+
```
Comment (mokagio) [inline WooCommerce/WooCommerceUITests/Flows/MockDataReader.swift]: Swift nitpick: This dictionary is all constants. If we extracted it as a `private static let` class property, it would be instantiated only once rather than every time this method is called.

I don't think it makes much difference from the point of view of the memory footprint, but I think it would make for "better looking" code.

File: WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift [hunk 2]
```diff
@@ -41,17 +41,32 @@ public final class SingleProductScreen: ScreenObject {
...
+            fatalError("Product Type \(productType) doesn't exist!")
+        }
     }
 
     public func verifyProductTypeScreenLoaded(productType: String) throws -> Self {
         let addPriceLabel = NSPredicate(format: "label == 'Add Price'")
         let inventoryLabel = NSPredicate(format: "label == 'Inventory'")
-        let productTypeLabel = NSPredicate(format: "label ==[c] '\(productType)'")
         let addVariationLabel = NSPredicate(format: "label == 'Add variations'")
+        let productTypeLabel = NSPredicate(format: "label ==[c] '\(productType)'")
```
Comment (mokagio) [inline WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift]: Just curious, why was this moved after the `addVariationLabel` definition? 🤷‍♂️

File: WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift [hunk 3]
```diff
@@ -41,17 +41,32 @@ public final class SingleProductScreen: ScreenObject {
...
+    public func verifyPublishedProductScreenLoaded(productType: String, productName: String) {
+        let priceLabel = NSPredicate(format: "label == 'Price'")
+        let addVariationLabel = NSPredicate(format: "label == 'Add variations'")
+
+        // common fields on a published product screen
         XCTAssertTrue(app.buttons["save-product-button"].exists)
         XCTAssertTrue(app.staticTexts["TIP"].exists)
         XCTAssertTrue(app.textViews[productName].exists)
+
+        // different product types displays different fields on the published product screen
```
Comment (mokagio) [inline WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift]:
```suggestion
        // different product types display different fields on the published product screen
```

File: WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift [hunk 4]
```diff
@@ -41,17 +41,32 @@ public final class SingleProductScreen: ScreenObject {
...
+        // this is to validate that the correct screens are displayed
+        switch productType {
+        case "physical", "virtual":
+            XCTAssertTrue(app.staticTexts.containing(priceLabel).firstMatch.exists)
+        case "variable":
+            XCTAssertTrue(app.staticTexts.containing(addVariationLabel).firstMatch.exists)
+        default:
+            fatalError("Product Type \(productType) doesn't exist!")
```
Comment (mokagio) [inline WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift]: Have you considered failing the test with `XCTFail` instead of crashing here?

Also, and this might be a change for a future PR, have you considered using an `enum` to encode the possible product types instead of a `String`? Doing so would make it impossible to run into this scenario. 

I'm a fan of the concept of **making undesirable state unrepresentable**. Evan Czaplicki, the author of the delightful programming language Elm, puts is better than I can in [this Appendix to the Elm documentation](https://https://guide.elm-lang.org/appendix/types_as_sets.html#who-cares), which I really recommend reading in full:

> So the whole point here is that ruling out invalid data makes your code shorter, simpler, and more reliable. By making sure the set of possible values in code exactly matches the set of valid values in real life, many problems just go away

Comment (jostnes) [reply WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift]: > Have you considered failing the test with XCTFail instead of crashing here?

i have not but i had a look and it does look to be better to use that to fail this test. updated in https://github.com/woocommerce/woocommerce-ios/pull/9441/commits/1b22d2f2bcb9b9bf340c7172bef3eb6927e70eb0

> have you considered using an enum to encode the possible product types instead of a String? Doing so would make it impossible to run into this scenario

good idea! and thanks for linking that doc, yeah the possibility of using the incorrect value is there (since the test is expecting a string value). i'll update this to an enum in a separate PR as there might be other functions using something similar to this.

File: WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift (hunk 2 above)
Comment (jostnes) [reply WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift]: ah ok, my bad, i was testing this out locally trying to see if i could move it out but it didn't work, when i added it i added it in a different order. i'll revert this so that it doesn't show as a code change. reverted in https://github.com/woocommerce/woocommerce-ios/pull/9441/commits/ecbb4b636a6e57c65da9de3e1d38b534adc8d4fa

File: WooCommerce/WooCommerceUITests/Flows/MockDataReader.swift (hunk 1 above)
Comment (jostnes) [reply WooCommerce/WooCommerceUITests/Flows/MockDataReader.swift]: updated in https://github.com/woocommerce/woocommerce-ios/pull/9441/commits/be54c931c0b6857b9d3681ae6226d2ddf96b05d2

File: WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift (hunk 3 above)
Comment (jostnes) [reply WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift]: updated in https://github.com/woocommerce/woocommerce-ios/pull/9441/commits/ecbb4b636a6e57c65da9de3e1d38b534adc8d4fa (also fixed another similar one)

File: WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift (hunk 4 above)
Comment (pachlava) [reply WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift]: It looks like 1b22d2f2bcb9b9bf340c7172bef3eb6927e70eb0 applies the `fatalError` ~> `XCTFail` change only for the L85, while on L65 the `fatalError` is still used.

File: WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift [hunk 5]
```diff
@@ -41,10 +41,25 @@ public final class SingleProductScreen: ScreenObject {
...
         XCTAssertTrue(app.staticTexts["TIP"].exists)
         XCTAssertTrue(app.textViews[productName].exists)
+
+        // different product types display different fields on the published product screen
+        // this is to validate that the correct screens are displayed
+        switch productType {
+        case "physical", "virtual":
+            XCTAssertTrue(app.staticTexts.containing(priceLabel).firstMatch.exists)
+        case "variable":
+            XCTAssertTrue(app.staticTexts.containing(addVariationLabel).firstMatch.exists)
```
Comment (pachlava) [inline WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift]: This will work too:

```suggestion
        case "physical", "virtual":
            XCTAssertTrue(app.staticTexts["Price"].exists)
        case "variable":
            XCTAssertTrue(app.staticTexts["Add variations"].exists)
```

Besides being shorter and allowing to remove two `NSPredicate`s at the function beginning, it's also stricter - if there will be more than one match, assertion will fail. With `firstMatch`, the assertion will not care it's there's more than one match.

Comment (pachlava) [inline WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift]: Maybe overthinking, but additionally checking for the absence of elements might improve robustness a bit (just an idea):

```suggestion
        case "physical", "virtual":
            XCTAssertTrue(app.staticTexts["Price"].exists)
            XCTAssertFalse(app.staticTexts["Add variations"].exists)
        case "variable":
            XCTAssertTrue(app.staticTexts["Add variations"].exists)
            XCTAssertFalse(app.staticTexts["Price"].exists)
```

**Review by pachlava — APPROVED** [review]
@jostnes 👋 

All the affected tests worked for me locally, CI is good too 👍 

I added several optimizing suggestions, and also one [comment](https://github.com/woocommerce/woocommerce-ios/pull/9441#discussion_r1165366100) about pending `fatalError` ~> `XCTFail` change.

I'm approving the PR, since the suggestions are nice-to-haves, IMO, but they are non-blocking ones.

File: WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift (hunk 4 above)
Comment (jostnes) [reply WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift]: missed this since it was existing code, but makes sense to update this in this PR. updated in https://github.com/woocommerce/woocommerce-ios/pull/9441/commits/44d8ab4ce93395b6af9387bc677a31c3ae7ee9c6

File: WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift (hunk 5 above)
Comment (jostnes) [reply WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift]: thanks for this suggestion! yeah makes sense and it reads better too. updated this and a similar function to use this way for exact match labels in https://github.com/woocommerce/woocommerce-ios/pull/9441/commits/8404f5679e27a4237a0e75a955b7bd6bfbfe7262

Comment (jostnes) [reply WooCommerce/UITestsFoundation/Screens/Products/SingleProductScreen.swift]: thanks for thinking about this too! but i think i'll keep only the true assert for now. adding the false assert will add extra run time to the test and i question the value that it adds with the added time since we're already asserting that the unique cells are visible on the screen
